File: goal-service/src/main/kotlin/de/rwth/swc/piggybank/goalservice/service/GoalService.kt
package de.rwth.swc.piggybank.goalservice.service

import de.rwth.swc.piggybank.goalservice.domain.Goal
import de.rwth.swc.piggybank.goalservice.domain.GoalStatus
import de.rwth.swc.piggybank.goalservice.domain.SavingsGoal
import de.rwth.swc.piggybank.goalservice.domain.SpendingLimitGoal
import de.rwth.swc.piggybank.goalservice.dto.CreateSavingsGoalRequest
import de.rwth.swc.piggybank.goalservice.dto.CreateSpendingLimitGoalRequest
import de.rwth.swc.piggybank.goalservice.dto.SavingsGoalResponse
import de.rwth.swc.piggybank.goalservice.dto.SpendingLimitGoalResponse
import java.util.UUID

/**
 * Service interface for goal operations.
 */
interface GoalService {
    /**
     * Creates a new spending limit goal.
     *
     * @param request The request containing the goal details
     * @return The created goal
     */
    fun createSpendingLimitGoal(request: CreateSpendingLimitGoalRequest): SpendingLimitGoalResponse

    /**
     * Creates a new savings goal.
     *
     * @param request The request containing the goal details
     * @return The created goal
     */
    fun createSavingsGoal(request: CreateSavingsGoalRequest): SavingsGoalResponse

    /**
     * Gets a goal by ID.
     *
     * @param id The ID of the goal
     * @return The goal, or null if not found
     */
    fun getGoalById(id: UUID): Goal?

    /**
     * Gets a spending limit goal by ID.
     *
     * @param id The ID of the goal
     * @return The goal, or null if not found
     */
    fun getSpendingLimitGoalById(id: UUID): SpendingLimitGoal?

    /**
     * Gets a savings goal by ID.
     *
     * @param id The ID of the goal
     * @return The goal, or null if not found
     */
    fun getSavingsGoalById(id: UUID): SavingsGoal?

    /**
     * Gets all goals for an account.
     *
     * @param accountId The ID of the account
     * @return A list of goals for the account
     */
    fun getGoalsByAccountId(accountId: String): List<Goal>

    /**
     * Gets all spending limit goals for an account.
     *
     * @param accountId The ID of the account
     * @return A list of spending limit goals for the account
     */
    fun getSpendingLimitGoalsByAccountId(accountId: String): List<SpendingLimitGoal>

    /**
     * Gets all savings goals for an account.
     *
     * @param accountId The ID of the account
     * @return A list of savings goals for the account
     */
    fun getSavingsGoalsByAccountId(accountId: String): List<SavingsGoal>

    /**
     * Gets all goals with a specific status.
     *
     * @param status The status of the goals
     * @return A list of goals with the status
     */
    fun getGoalsByStatus(status: GoalStatus): List<Goal>

    /**
     * Gets all spending limit goals for a specific category.
     *
     * @param category The category of the goals
     * @return A list of spending limit goals for the category
     */
    fun getSpendingLimitGoalsByCategory(category: String): List<SpendingLimitGoal>

    /**
     * Updates the status of a goal.
     *
     * @param id The ID of the goal
     * @param status The new status of the goal
     * @return The updated goal, or null if not found
     */
    fun updateGoalStatus(id: UUID, status: GoalStatus): Goal?

    /**
     * Deletes a goal.
     *
     * @param id The ID of the goal
     * @return true if the goal was deleted, false otherwise
     */
    fun deleteGoal(id: UUID): Boolean
}

File: piggybank-ui/src/services/goalService.ts
import axios from 'axios';
import { 
  Goal, 
  SavingsGoal, 
  SpendingLimitGoal, 
  CreateSavingsGoalRequest, 
  CreateSpendingLimitGoalRequest, 
  GoalStatus 
} from '../types';

const GOALS_API_URL = '/api/goals';

// Goal functions
export const getGoalById = async (id: string): Promise<Goal> => {
  const response = await axios.get<Goal>(`${GOALS_API_URL}/${id}`);
  return response.data;
};

export const getGoalsByAccountId = async (accountId: string): Promise<Goal[]> => {
  const response = await axios.get<Goal[]>(`${GOALS_API_URL}/account/${accountId}`);
  return response.data;
};

export const getGoalsByStatus = async (status: GoalStatus): Promise<Goal[]> => {
  const response = await axios.get<Goal[]>(`${GOALS_API_URL}/status/${status}`);
  return response.data;
};

// Savings goal functions
export const createSavingsGoal = async (
  request: CreateSavingsGoalRequest
): Promise<SavingsGoal> => {
  // Transform the request to match the backend's expected format
  const transformedRequest = {
    name: request.name,
    description: request.description,
    startDate: request.startDate,
    endDate: request.endDate,
    accountId: request.accountId,
    targetAmount: request.targetAmount.value,
    currencyCode: request.targetAmount.currencyCode
  };

  const response = await axios.post<SavingsGoal>(`${GOALS_API_URL}/savings`, transformedRequest);
  return response.data;
};

export const getSavingsGoalById = async (id: string): Promise<SavingsGoal> => {
  const response = await axios.get<any>(`${GOALS_API_URL}/savings/${id}`);
  // Transform the response to match the frontend's expected format
  const data = response.data;
  return {
    ...data,
    targetAmount: {
      value: data.targetAmount,
      currencyCode: data.currencyCode
    },
    currentAmount: {
      value: data.currentAmount,
      currencyCode: data.currencyCode
    }
  };
};

export const getSavingsGoalsByAccountId = async (accountId: string): Promise<SavingsGoal[]> => {
  const response = await axios.get<any[]>(`${GOALS_API_URL}/savings/account/${accountId}`);
  // Transform the response to match the frontend's expected format
  return response.data.map(data => ({
    ...data,
    targetAmount: {
      value: data.targetAmount,
      currencyCode: data.currencyCode
    },
    currentAmount: {
      value: data.currentAmount,
      currencyCode: data.currencyCode
    }
  }));
};

// Spending limit goal functions
export const createSpendingLimitGoal = async (
  request: CreateSpendingLimitGoalRequest
): Promise<SpendingLimitGoal> => {
  // Transform the request to match the backend's expected format
  const transformedRequest = {
    name: request.name,
    description: request.description,
    startDate: request.startDate,
    endDate: request.endDate,
    accountId: request.accountId,
    limit: request.limit.value,
    currencyCode: request.limit.currencyCode,
    category: request.category
  };

  const response = await axios.post<SpendingLimitGoal>(`${GOALS_API_URL}/spending-limit`, transformedRequest);
  return response.data;
};

export const getSpendingLimitGoalById = async (id: string): Promise<SpendingLimitGoal> => {
  const response = await axios.get<any>(`${GOALS_API_URL}/spending-limit/${id}`);
  // Transform the response to match the frontend's expected format
  const data = response.data;
  return {
    ...data,
    limit: {
      value: data.limit,
      currencyCode: data.currencyCode
    },
    currentSpending: {
      value: data.currentSpending,
      currencyCode: data.currencyCode
    }
  };
};

export const getSpendingLimitGoalsByAccountId = async (accountId: string): Promise<SpendingLimitGoal[]> => {
  const response = await axios.get<any[]>(`${GOALS_API_URL}/spending-limit/account/${accountId}`);
  // Transform the response to match the frontend's expected format
  return response.data.map(data => ({
    ...data,
    limit: {
      value: data.limit,
      currencyCode: data.currencyCode
    },
    currentSpending: {
      value: data.currentSpending,
      currencyCode: data.currencyCode
    }
  }));
};

export const getSpendingLimitGoalsByCategory = async (category: string): Promise<SpendingLimitGoal[]> => {
  const response = await axios.get<any[]>(`${GOALS_API_URL}/spending-limit/category/${category}`);
  // Transform the response to match the frontend's expected format
  return response.data.map(data => ({
    ...data,
    limit: {
      value: data.limit,
      currencyCode: data.currencyCode
    },
    currentSpending: {
      value: data.currentSpending,
      currencyCode: data.currencyCode
    }
  }));
};

// Common goal functions
export const updateGoalStatus = async (id: string, status: GoalStatus): Promise<Goal> => {
  const response = await axios.put<Goal>(`${GOALS_API_URL}/${id}/status/${status}`);
  return response.data;
};

export const deleteGoal = async (id: string): Promise<void> => {
  await axios.delete(`${GOALS_API_URL}/${id}`);
};
